### PR TITLE
Feature/ZING-1117

### PR DESCRIPTION
**Pull request by fraibacas**
Changes to forward events to Zenoss Cloud. Events are forwarded in a zep post indexing plugin.

This PR allows to forward events to either GCP pubsub or an emulator.

Sample config to forward to GCP:
```
zep.zing.enabled=true
zep.zing.use_emulator=false
zep.zing.tenant=ACME
zep.zing.source=cz1
zep.zing.project=zing-dev-197522
zep.zing.topic=event-in
zep.zing.credentials=/tmp/creds.json
```

Sample config to forward to emulator:
```
zep.zing.enabled=true
zep.zing.use_emulator=true
zep.zing.tenant=ACME
zep.zing.source=cz1
zep.zing.project=zenoss-zing
zep.zing.topic=eventos-in
zep.zing.emulator_url=paco.zenoss.eng:8085
```

To disable event forwarding (needed for 6.X)
```
zep.zing.enabled=false
```

TODOS:
- Write tests
- Figure out what to do when pubsub is not available or returns an error
- Instrumentation. Expose instrumentation in zep CC service
- Properly expose configuration in CC

### CODE REVIEW

File: core/src/main/java/org/zenoss/zep/zing/impl/ZingPublisher.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
+/*****************************************************************************
+ *
+ * Copyright (C) Zenoss, Inc. 2018, all rights reserved.
+ *
+ * This content is made available according to terms specified in
+ * License.zenoss under the directory where your Zenoss product is installed.
+ *
+ ****************************************************************************/
+
+package org.zenoss.zep.zing.impl;
```
Comment (bharatZen) [inline core/src/main/java/org/zenoss/zep/zing/impl/ZingPublisher.java]: Should this be one level up?

File: core/src/main/resources/zeneventserver.conf [hunk 2]
```diff
@@ -248,3 +248,13 @@ zep.redis.port=6379
 #zep.lucene.archive.status=READER
 #zep.lucene.archive.async.updates=false
 #zep.lucene.archive.honor.deletes=true
+
+# Zenoss Cloud configuration, disabled by default
+zep.zing.enabled=true
+zep.zing.use_emulator=true
+zep.zing.tenant=paco
```
Comment (bharatZen) [inline core/src/main/resources/zeneventserver.conf]: How will the tenant and source be coordinated with the overall CZ?

File: core/src/main/java/org/zenoss/zep/zing/ZingEvent.java [hunk 3]
```diff
@@ -0,0 +1,421 @@
...
+        Event evt = null;
+        if (this.isValid()) {
+            Event.Builder b = Event.newBuilder();
+            b.setTenant(this.tenant);
+            // The timestamp is set to the event's lastSeen (raw event occurenceTime) which
+            // can be confusing when sending event updates (state change, detail/log update, etc)
+            //      - lastSeen will always contain the event's last occurence time
+            //      - updateTime will always contain the last time the event was updated
+            //        (new occurence or event updates)
+            b.setTimestamp(this.occurrenceTime);
```
[the remark this answers is not in the document]
Comment (fraibacas) [reply core/src/main/java/org/zenoss/zep/zing/ZingEvent.java]: lastSeen should match that

File: core/src/main/resources/zeneventserver.conf (hunk 2 above)
Comment (fraibacas) [reply core/src/main/resources/zeneventserver.conf]: This is one of the items in my TODO list. I will address it in a future PR.

File: core/src/main/java/org/zenoss/zep/zing/ZingPublisher.java [hunk 4]
```diff
@@ -19,19 +19,40 @@
 import com.google.pubsub.v1.PubsubMessage;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.zenoss.zep.zing.ZingConfig;
-import org.zenoss.zep.zing.ZingEvent;
 import org.zenoss.zing.proto.event.Event;
 
-public abstract class ZingPublisher {
+public class ZingPublisher {
```
Comment (bharatZen) [inline core/src/main/java/org/zenoss/zep/zing/ZingPublisher.java]: Are you sure you don't want this to be abstract? will it work if instantiated at this level (i.e. not via a subclass)? If not, keep it as an abstract class

**Pull request merged**